### ISSUE TITLE
Bug: [prefer-nullish-coalescing] <Misidentification of top-level object constructors>

### ISSUE THREAD

**Issue by linbudu599** [typescript-eslint/typescript-eslint#6543]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.9.3&sourceType=module&code=GYVwdgxgLglg9mABMOcAUMwAcRQFyLgDWYcA7mAJQEBGqANgKYCGSAPoWACaPCaNdEAbwBQiRACdGUEBKTBm9AM6MA3CIC+IkQCEGLMGhToAjJURsOxtACZKlIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WlqYSNkAC1pkA9gEMkyMswDm6KL2jjokcGAC+ILUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
function foo(input: unknown): boolean | undefined {
  return false;
}

Boolean(foo(1) || foo(2))
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/prefer-nullish-coalescing": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

`||` syntax inside Boolean constructor should be regarded as conditional tests, so `ignoreConditionalTests` should also works for this. 

### Actual Result

Should not report error in repro code.

**Comment by bradzacher** [comment]
> syntax inside Boolean constructor should be regarded as conditional tests

Why should it be regarded as a conditional test?
What's the difference between `Boolean()` and `someFunctionThatReturnsBoolean`?

**Comment by suienzan** [comment]
Seems unrelated to `Boolean()`. Likely caused by `boolean | undefined`.

```ts
type User = {
  admin?: boolean;
};
const foo: User = { admin: false};
const bar: User = { admin: true};
const hasAdmin = foo.admin || bar.admin; // true, report error here
const wrongHasAdmin = foo.admin ?? bar.admin; // false
```

[Playground Link](https://typescript-eslint.io/play/#ts=5.0.2&sourceType=module&code=C4TwDgpgBAqgzhATlAvFA3gKClAhgEwFsBLAOwH4AuKAIwHs6AbCXUgbkwF8OBjO0uMCgAzBtXhJUGPETLVhuRgijdMfAUJq5E4hMjToZJUtWCIArtFXrBUABa44AQVmkpougDoCxqAB8-Wm1vVzYoAHpwqDNLABooRAgwOkQhJEQU+yQINX5bAHcM0gBzAAlHF180DxDfcnIgxFqyMMiRRQQgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WlqYSNkAC1pkA9gEMkyMswDm6KL2jjokcGAC+ILUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)
